[clusteragent/admission/sidecar] Add support for APM injection on EKS Fargate

davidor opened the pull request:
### What does this PR do?

Modifies the sidecar admission controller webhook so that APM library injection works when deployed on EKS Fargate.

It was not working because the "config" webhook creates a "HostPath" volume to mount the APM socket and that doesn't work on EKS Fargate.

This PR deletes the volume and volume mounts created by the "config" webhook when using EKS Fargate and creates an "emptyDir" volume instead. It also sets the needed envs so that the sidecar Agent receives traces via UDS.


### Describe how to test/QA your changes

Deploy the Agent on EKS Fargate. You need to enable the admission controller with APM and sidecar injection. Here's the helm chart that I used:
```
datadog:
  processAgent:
    processCollection: true
  env:
    - name: DD_LANGUAGE_DETECTION_ENABLED
      value: "true"
    - name: DD_PROCESS_CONFIG_PROCESS_COLLECTION_ENABLED
      value: "true"
clusterAgent:
  admissionController:
    enabled: true
  env:
    - name: DD_LANGUAGE_DETECTION_ENABLED
      value: "true"
    - name: DD_APM_INSTRUMENTATION_ENABLED
      value: "true"
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_ENABLED
      value: "true"
    - name: DD_ADMISSION_CONTROLLER_AGENT_SIDECAR_PROVIDER
      value: "fargate"
```

You'll also need an app with profiling enabled. For testing purposes I've used this:
```
import ddtrace.profiling.auto
import time

while True:
  time.sleep(60)
```

And finally deploy the app in the Fargate namespace:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: my-python-app
spec:
  replicas: 1
  selector:
    matchLabels:
      app: python
  template:
    metadata:
      labels:
        app: python
        admission.datadoghq.com/enabled: "true"
        agent.datadoghq.com/sidecar: "fargate"
      annotations:
        admission.datadoghq.com/python-lib.version: "v2.3.0"
    spec:
      serviceAccountName: datadog-agent  # Make sure this matches your service account
      shareProcessNamespace: true
      containers:
      - name: my-python-app
        image: YOUR_IMAGE_HERE
        command: ["python", "./test.py"] # Make sure this matches what's in your image
```

Remember to create the needed RBACs for Fargate as described here: https://docs.datadoghq.com/integrations/eks_fargate/#aws-eks-fargate-rbac

You should see APM traces in the Datadog app for your service and they should include the container tags such as `container_name`, `image_name`, `image_tag`, etc.

davidor committed:
[clusteragent/admission/sidecar] Add support for APM injection on EKS Fargate

davidor committed:
Add release note

clamoriniere commented on pkg/clusteragent/admission/mutate/agent_sidecar/agent_sidecar_test.go:
```diff
@@ -144,20 +146,58 @@ func TestInjectAgentSidecar(t *testing.T) {
+
+				sidecar.VolumeMounts = []corev1.VolumeMount{
+					{
+						Name:      "apmsocket",
```
can we also mount the dogstatsd socket?

clamoriniere commented on pkg/clusteragent/admission/mutate/agent_sidecar/providers.go:
```diff
@@ -21,6 +23,10 @@ import (
 //                            //
 ////////////////////////////////
 
+const apmSocketDir = "/var/run/datadog"
```
we should also mount the dogstatsd socket.
maybe we don't need to volume but one that container both socket

davidor committed:
[clusteragent/admission/mutate] Don't try to patch pod if mutate func returns err

davidor committed:
[clusteragent/admission/sidecar] Configure DogStatsD socket on EKS Fargate

davidor commented:
@clamoriniere I rebased on main and added a couple of new commits since you reviewed. The first one fixes a bug introduced in a previous PR (https://github.com/DataDog/datadog-agent/pull/23770) and the second one configures DogStatsD as you suggested.

davidor committed:
Fix linter issue

davidor commented:
Rebased on main

davidor commented:
/merge
